File: api/src/main/java/club/yunzhi/smartcommunity/entity/Role.java
package club.yunzhi.smartcommunity.entity;

import club.yunzhi.smartcommunity.enums.RoleType;
import com.fasterxml.jackson.annotation.JsonView;
import org.hibernate.annotations.SQLDelete;
;
import org.hibernate.annotations.Where;
import org.springframework.security.core.GrantedAuthority;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

/**
 * 角色.
 */
@Entity
@SQLDelete(sql = "update `role` set deleted = 1 where id = ?")
public class Role extends SoftDeleteEntity implements GrantedAuthority {
  @Column(nullable = false, unique = true)
  private String name;

  @Column(nullable = false, unique = true)
  private String value;

  private int weight;

  /**
   * 该角色的子角色
   */
  @ManyToMany
  @Where(clause = "deleted = false")
  @JsonView(ChildrenJsonView.class)
  private List<Role> children = new ArrayList<>();

  /**
   * 是否系统内置角色
   */
  private Boolean systemed = false;

  public Role() {
  }

  public Role(String name, String value, int weight, Boolean systemed) {
    this.name = name;
    this.value = value;
    this.weight = weight;
    this.systemed = systemed;
  }

  public Role(RoleType roleType, int weight, Boolean systemed) {
    this(roleType.getName(), roleType.getValue(), weight, systemed);
  }

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public String getValue() {
    return value;
  }

  public void setValue(String value) {
    this.value = value;
  }

  public int getWeight() {
    return weight;
  }

  public void setWeight(int weight) {
    this.weight = weight;
  }


  public Boolean getSystemed() {
    return systemed;
  }

  public void setSystemed(Boolean systemed) {
    this.systemed = systemed;
  }

  public List<Role> getChildren() {
    return children;
  }

  public void setChildren(List<Role> children) {
    this.children = children;
  }

  @Override
  public String getAuthority() {
    return this.value;
  }

  public interface ChildrenJsonView {
  }
}

File: web/projects/lib/src/entity/role.ts
/**
 * 角色.
 */
export class Role {
  children: Role[];
  /**
   * id
   */
  id: number;
  /**
   * 名称
   */
  name: string;
  /**
   * 是否系统内置角色
   */
  systemed: boolean;
  /**
   * 值
   * 用于前后台对接，该值具有唯一性
   */
  value: string;
  /**
   * 权重
   */
  weight: number;

  constructor(data = {} as {
    id?: number,
    name?: string,
    value?: string,
    weight?: number,
    systemed?: boolean,
    children?: Role[]
  }) {
    if (data) {
      this.id = data.id;
      this.name = data.name;
      this.value = data.value;
      this.weight = data.weight;
      this.systemed = data.systemed;
      if (data.children && Array.isArray(data.children)) {
        this.children = data.children.map(role => role instanceof Role ? role : new Role(role));
      }
    }
  }
}
